### DISCUSSION TITLE
Can't push to forked repository on the original repository's pull request with GITHUB_TOKEN

### DISCUSSION THREAD

**Discussion by kou** [community/community#26904] [Actions]
<p>TL;DR:</p>
<p>Curretly, we can’t push to the forked repository with GITHUB_TOKEN in worlflow that is triggered by pull request for the original repository from the forked repository. Is it expected behavior?</p>
<p>FYI: We can push to the original repository with GITHUB_TOKEN in workflow that is triggered by pull request for the original repository from the original repository.</p>
<p>Details:</p>
<p>I want to create a workflow that prepares GitHub Pages to confirm pull requested change automatically. Here is a scenario I expect:</p>
<ol>
<li>Contributor forks the original repository to submit a pull request</li>
<li>Contributor creates a topic branch on the contributor’s forked repository</li>
<li>Contributor submits a pull request for the topic branch to the original repository</li>
<li>GitHub Actions is triggered on the original repository</li>
<li>The triggered GitHub Actions builds site and pushes the built site to the “gh-pages” branch on the contributor’s forked repository with GITHUB_TOKEN</li>
<li>Contributor and maintainers can confirm the result of the change in the topic branch at <a>https://$</a>{contributor_id}.github.io/${repository_name} [1]</li>
</ol>
<p>[1] I know that GitHub Actions and GitHub Pages integration have a problem: <a href="https://github.community/t/github-action-not-triggering-gh-pages-upon-push/16096">https://github.community/t5/GitHub-Actions/Github-action-not-triggering-gh-pages-upon-push/td-p/26869</a></p>
<p>In this scenario, “git push forked gh-pages” in 5. is failed by “Permission to ${forked-repository} denied to github-actions[bot]”.</p>
<p>This scenario works well when we use the same repository instead of forked repository:</p>
<ol>
<li>Contributor creates a topic branch on the original repository</li>
<li>Contributor submits a pull request for the topic branch to the original repository</li>
<li>GitHub Actions is triggered on the original repository</li>
<li>The triggered GitHub Actions builds site and pushes the built site to the “gh-pages” branch on the original repository with GITHUB_TOKEN</li>
<li>Contributor and maintainers can confirm the result of the change in the topic branch at <a>https://$</a>{original_repository_owner}.github.io/${repository_name} [1]</li>
</ol>
<p>Here is a workflow file I created: <a href="https://github.com/apache/arrow-site/blob/ccdb39a0e32802bbb6c94961ac5f4deea730a932/.github/workflows/deploy.yml#L59-L77" rel="nofollow noopener">https://github.com/apache/arrow-site/blob/ccdb39a0e32802bbb6c94961ac5f4deea730a932/.github/workflows/deploy.yml#L59-L77</a></p>
<p>It doesn’t work on pull request on the original repository: <a href="https://github.com/apache/arrow-site/pull/26/checks#step:8:40" rel="nofollow noopener">https://github.com/apache/arrow-site/pull/26/checks#step:8:40</a></p>
<p>It works on pull request on the same repository: <a href="https://github.com/kou/arrow-site/runs/227573664#step:8:40" rel="nofollow noopener">https://github.com/kou/arrow-site/runs/227573664#step:8:40</a></p>

**Comment by chrispat** [comment, marked as answer]
<p>This is expected behavior, as mentioned here <a href="https://help.github.com/en/github/automating-your-workflow-with-github-actions/virtual-environments-for-github-actions#token-permissions" rel="nofollow noopener">https://help.github.com/en/github/automating-your-workflow-with-github-actions/virtual-environments-for-github-actions#token-permissions</a> when running a pull_request form a fork the token can’t push contents to either repo.  This is to prevent melicious actors from using actions to poision upstrem or downstream repos.</p>

**Comment by cclauss** [comment]
<p>Configuration-less code formatters like gofmt, black,  standard would be good to be able to autorun on inbound pull requests (especially during Hacktoberfest when many green developers jump on GitHub).  I can understand the security concerns of allowing all Actions to write back to the inbound pull requests.</p>
<p>It would be cool if GitHub could figure out a non-Actions based solution for running configurationless formatters in inbound pull requests.</p>

**Comment by kou** [comment]
<p>Thanks. I understand.</p>

**Comment by brettcannon** [comment]
<p>In case anyone is looking for an up-to-date URL to the token permissions section of the docs it’s <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token" rel="nofollow noopener">https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token</a>.</p>

**Comment by glyph** [comment]
<p>I’m still a bit confused as to whether it’s possible to supply a secondary token that isn’t GITHUB_TOKEN which will be able to push back to PR branches.  Is that possible to set up, and if so, what permissions does it need?</p>

**Comment by kingthorin** [comment]
<p>Since PRs default to allow updates from maintainers then you could create a maintainer bot account and use a PAT to commit to PR forks.</p>
<p>I’m not sure about the specific permissions needed.</p>
